src/torrent.cpp

```diff
@@ -7429,23 +7429,52 @@ bool is_downloading_state(int const st)
 		if (lhs->on_parole() != rhs->on_parole())
 			return lhs->on_parole();
 
-		// prefer to disconnect peers that send data at a lower rate
+		// prefer to disconnect peers that send data at a lower rate over those waiting to download
+		// the assumption is that a connection waiting to start downloading has the same potential 
+		// to be faster than the current slowest downloading connection than a new connection that 
+		// replaces a waiting connection - and disconnecting a waiting connection starts a new wait
+		// however ... to avoid torrents which wait forever there needs to be a maximum wait time
+		// after which a waiting connection is disconnected
 		std::int64_t lhs_transferred = lhs->statistics().total_payload_download();
 		std::int64_t rhs_transferred = rhs->statistics().total_payload_download();
+		bool const is_waiting = lhs_transferred * rhs_transferred == 0;
 
+		// for this commit, the maximum time is a constant - but it could be replaced by a new 
+		// setting peer_turnover_maximum_wait in a future commit
+		std::int64_t const peer_turnover_maximum_wait = 60;
 		time_point const now = aux::time_now();
 		std::int64_t const lhs_time_connected = total_seconds(now - lhs->connected_time());
 		std::int64_t const rhs_time_connected = total_seconds(now - rhs->connected_time());
+		
+		// if one connection is waiting
+		if ((lhs_transferred != rhs_transferred) && (is_waiting))
+		{
+			// One of lhs_transferred & rhs_transferred is zero - get connected time for waiting connection
+			std::int64_t const waiting_time = 
```

Comment:
this computation seems a bit awkward. Is there a good reason to do this rather than the straigh-forward:

```
auto const waiting_type = (lhs_transferred == 0) ? lhs_time_connected : rhs_time_connected;
```

I think that's easier for a human to interpret and you don't need any multiplication or division.

Reply:
Agreed.